File: backend/src/main/java/feup/sdle/ShoppingList.java
package feup.sdle;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import com.fasterxml.jackson.annotation.JsonTypeName;
import com.google.protobuf.ByteString;
import com.google.protobuf.MapEntry;
import feup.sdle.Document;
import feup.sdle.cluster.NodeIdentifier;
import feup.sdle.crdts.AWMap;
import feup.sdle.crdts.DottedValue;
import feup.sdle.crdts.ShoppingListItem;
import feup.sdle.message.DocumentProto;
import feup.sdle.message.DottedValueProto;
import feup.sdle.message.NodeIdentifierMessage;
import feup.sdle.utils.Color;

import java.nio.channels.InterruptedByTimeoutException;
import java.util.*;

@JsonTypeInfo(
        use = JsonTypeInfo.Id.NAME,
        include = JsonTypeInfo.As.PROPERTY
)

/**
    * This is a CRDT set that will need to mantain a list of removed items, so in cases of a concurrent edit
    * the behaviour of the counters will not be wrong
*/
@JsonTypeName("shoppingList")
public class ShoppingList implements Document {
    @JsonProperty("id")
    private String id;
    // The String key will never change. This is an id of the shopping list and not the name of the list itself.
    @JsonProperty("items")
    private AWMap<String, ShoppingListItem> items;
    @JsonProperty("localIdentifier")
    private int localIdentifier;
    @JsonProperty("removedCounters")
    private HashMap<String, DottedValue<Integer, Integer, Integer>> removedCounters;

    @JsonCreator
    public ShoppingList(@JsonProperty("id") String id, @JsonProperty("items") AWMap<String, ShoppingListItem> items,
                        @JsonProperty("localIdentifier") int localIdentifier, @JsonProperty("removedCounters") HashMap<String, DottedValue<Integer, Integer, Integer>> removedCounters) {
        this.id = id;
        this.items = items;
        this.localIdentifier = localIdentifier;
        this.removedCounters = removedCounters;
    }

    public ShoppingList(int localIdentifier, String id) {
        this.items = new AWMap<>(localIdentifier);
        this.localIdentifier = localIdentifier;
        this.removedCounters = new HashMap<>();
        this.id = id;
    }

    public ShoppingList(int localIdentifier) {
        this.items = new AWMap<>(localIdentifier);
        this.localIdentifier = localIdentifier;
        this.removedCounters = new HashMap<>();
        this.id = UUID.randomUUID().toString();
    }

    public void setNodeIdentifier(int nodeIdentifier) {
        this.localIdentifier = nodeIdentifier;
        this.items.setNodeIdentifier(nodeIdentifier);
    }

    public HashMap<String, DottedValue<Integer, Integer, Integer>> getRemovedCounters() {
        return removedCounters;
    }

    public void setItems(AWMap<String, ShoppingListItem> items) {
        this.items = items;
    }

    public void setRemovedCounters(HashMap<String, DottedValue<Integer, Integer, Integer>> removedCounters) {
        this.removedCounters = removedCounters;
    }

    public void addItem(String key, String name, int quantity) {
        this.items.add(key, new ShoppingListItem(key, this.localIdentifier, name, quantity));
    }

    public void remove(String key) {
        DottedValue<Integer, Integer, ShoppingListItem> item = this.items.getValue(key);

        DottedValue<Integer, Integer, Integer> removed = this.removedCounters.get(key);
        if(removed != null) {
            this.removedCounters.put(key, new DottedValue<>(item.identifier(), item.event() + 1, item.value().getQuantity() + removed.value()));
            item.value().updateQuantity(-(item.value().getQuantity() + removed.value()));
        } else {
            this.removedCounters.put(key, new DottedValue<>(item.identifier(), item.event() + 1, item.value().getQuantity()));
            item.value().updateQuantity(-(item.value().getQuantity()));
        }

        this.items.remove(key);
    }

     /**
      * When merging we have to see if there is a dotted value with a dot of a remove event that we have after the remove.
      * If so, we have to do change the value of the current counter of the key by decrementing its value.
      *
      * For example, if someone removes the item with the original value 1 but the other one adds 2 to the quantity, the item should remain with value 1.vgit
      */
     @Override
     public void merge(Document otherDocument) {
        if(!(otherDocument instanceof ShoppingList other)) {
            return;
        }

        var modifiedItems = this.getModifiedItems(other.getItems().getValues());

        Optional<Integer> latestOtherDot = this.items.getDotContext().latestReplicaDot(other.localIdentifier);

        this.items.merge(other.items);

        for(Map.Entry<String, DottedValue<Integer, Integer, Integer>> entry: other.removedCounters.entrySet()) {
            if(entry.getValue().identifier() == other.localIdentifier
                    && entry.getValue().event() > latestOtherDot.orElse(0)
                    && this.items.getValue(entry.getKey()) != null) {
                this.removedCounters.put(entry.getKey(), entry.getValue());
                var thisValue = this.items.getValue(entry.getKey());
                var otherValue = other.getItems().getValue(entry.getKey());

                if(thisValue != null && otherValue != null && (!thisValue.value().getCounter().isConcurrent(otherValue.value().getCounter()))) {
                    this.items.remove(entry.getKey());
                    this.removedCounters.put(entry.getKey(), entry.getValue());
                } else {
                    this.items.getValue(entry.getKey()).value().updateQuantity(-entry.getValue().value());
                }
            }
        }

         /*for(Map.Entry<String, DottedValue<Integer, Integer, Integer>> entry: this.removedCounters.entrySet()) {
             if(this.items.getValue(entry.getKey()).value().getCounter().isConcurrent(other.getItems().getValue(entry.getKey()).value().getCounter())) {
                 this.removedCounters.remove(entry.getKey());
             }
         }*/


        /*for (Map.Entry<String, DottedValue<Integer, Integer, Integer>> entry : this.removedCounters.entrySet()) {
            if (this.getItems().getValue(entry.getKey()) != null && !modifiedItems.contains(entry.getKey())) {
                this.getItems().remove(entry.getKey());
            }
        }*/
    }

    private List<String> getModifiedItems(HashMap<String, DottedValue<Integer, Integer, ShoppingListItem>> otherListValues) {
         List<String> modifiedItems = new ArrayList<>();

         for (Map.Entry<String, DottedValue<Integer, Integer, ShoppingListItem>> entry : otherListValues.entrySet()) {
             var localValue = this.items.getValue(entry.getKey());

             if (localValue != null && localValue.value().getQuantity() < entry.getValue().value().getQuantity()) {
                 modifiedItems.add(entry.getKey());
             }
         }

         return modifiedItems;
    }

    public AWMap<String, ShoppingListItem> getItems() {
        return this.items;
    }

    @Override
    public String getId() {
        return this.id;
    }

    @Override
    public void setId(String id) {
        this.id = id;
    }

    @Override
    public ByteString toProtoBuf() {
        return this.toMessage().toByteString();
    }

    @Override
    public DocumentProto.Document toMessage() {
        var builder = DocumentProto.ShoppingList.newBuilder()
                .setItems(this.items.toMessageProto())
                .setLocalIdentifier(this.localIdentifier);

        HashMap<String, DottedValueProto.DottedValue> protoEntries = new HashMap<>();

        for (Map.Entry<String, DottedValue<Integer, Integer, Integer>> entry : this.removedCounters.entrySet()) {
            protoEntries.put(entry.getKey(), entry.getValue().toMessageDottedValue());
        }

        var shoppingList = builder.putAllRemovedCounters(protoEntries).build();

        var documentMessage = DocumentProto.Document.newBuilder()
                .setShoppingList(shoppingList)
                .setKey(this.id)
                .build();

        return documentMessage;
    }

    public static ShoppingList fromMessage(DocumentProto.Document document) {
        var msgShoppingList = document.getShoppingList();
        ShoppingList shoppingList = new ShoppingList(msgShoppingList.getLocalIdentifier());
        shoppingList.setId(document.getKey());
        shoppingList.setItems(AWMap.fromMessageAWMap(msgShoppingList.getItems()));

        HashMap<String , DottedValue<Integer, Integer, Integer>> removedCounters = new HashMap<>();
        for (Map.Entry<String, DottedValueProto.DottedValue> entry : msgShoppingList.getRemovedCountersMap().entrySet()) {
            removedCounters.put(entry.getKey(), DottedValue.fromMessageDottedValueInt(entry.getValue()));
        }

        shoppingList.setRemovedCounters(removedCounters);
        return shoppingList;
    }
}

File: frontend/src/lib/crdts/ShoppingList.ts
import { AWMap } from "./AWMap";
import { DottedValue } from "./DottedValue";
import { NodeIdentifier } from "../p2p/NodeIdentifier";
import { ShoppingListItem } from "./ShoppingListItem";
import { HashRing } from "../p2p/HashRing";

export class ShoppingList {
    private id: string;
    private lastItemId: number;
    private items: AWMap<string, ShoppingListItem>;
    private localIdentifier: number;
    private removedCounters: Map<string, DottedValue<number, number, number>>;

    constructor(localIdentifier: number, id?: string) {
        this.localIdentifier = localIdentifier;
        this.id = id || crypto.randomUUID();
        this.lastItemId = 0;
        this.items = new AWMap<string, ShoppingListItem>(this.localIdentifier);
        this.removedCounters = new Map<string, DottedValue<number, number, number>>();
    }

    public toSerializable(local: boolean = true) {

        // serialize removedCounters
        let map = null;

        if (local) {
            map = new Map();
        }
        else {
            map = {};
        }

        console.log("CURRENT REMOVE COUNTERS: ", this.removedCounters);

        if (this.removedCounters) {
            for (const [k,v] of this.removedCounters) {
                if (local) {
                    map.set(k, new DottedValue(v.identifier, v.event, v.value));
                }
                else {
                    map[k] = new DottedValue(v.identifier, v.event, v.value);
                }
            }
        }

        return {
            "type": "shoppingList",
            "id": this.id,
            "localIdentifier": this.localIdentifier,
            "items": this.items.toSerializable(local),
            "removedCounters": map
        }
    }

    public getLocalIdentifier() {
        return this.localIdentifier;
    }

    public setNodeIdentifier(identifier: number): void {
        this.localIdentifier = identifier;
    }

    public setItems(items: AWMap<string, ShoppingListItem>): void {
        this.items = items;
    }

    public setRemovedCounters(removedCounters: Map<string, DottedValue<number, number, number>>): void {
        this.removedCounters = removedCounters;
    }

    public addItem(key: string, name: string, quantity: number): void {
        this.items.add(key, new ShoppingListItem(key, this.localIdentifier, name, quantity));
    }

    public remove(key: string): void {
        const item = this.items.getValue(key);
        if (!item) return;

        const removed = this.removedCounters.get(key);
        if (removed) {
            this.removedCounters.set(
                key,
                new DottedValue<number, number, number>(
                    item.identifier,
                    item.event + 1,
                    item.value.getQuantity() + removed.value
                )
            );
        } else {
            this.removedCounters.set(
                key,
                new DottedValue<number, number, number>(
                    item.identifier,
                    item.event + 1,
                    item.value.getQuantity()
                )
            );
        }

        this.items.remove(key);
    }

    public merge(other: ShoppingList): void {
        
        console.log("MERGING SHOPPING LIST: ", other);

        const latestOtherDot = this.items.getDotContext().latestReplicaDot(other.localIdentifier);

        this.items.merge(other.items);

        if (!latestOtherDot) {
            console.warn("No latest other dot!");
            return;
        }

        for (const [key, dottedValue] of other.removedCounters.entries()) {
            if (
                dottedValue.identifier === other.localIdentifier.getId() &&
                dottedValue.event > latestOtherDot
            ) {
                const currentValue = this.items.getValue(key);
                if (currentValue) {
                    currentValue.value.updateQuantity(-dottedValue.value);
                }
            }

            this.removedCounters.set(
                key,
                dottedValue
            );
        }
    }

    public getItems(): AWMap<string, ShoppingListItem> {
        return this.items;
    }

    public getId(): string {
        return this.id;
    }

    public setId(id: string): void {
        this.id = id;
    }

    public toProtoBuf(): ShoppingListProto {
        const removedCountersSerialized: { [key: string]: any } = {};

        this.removedCounters.forEach((value, key) => {
            removedCountersSerialized[key] = value.toMessageDottedValue();
        });

        return {
            id: this.id,
            items: this.items.toMessageProto(),
            localIdentifier: this.localIdentifier.toMessageNodeIdentifier(),
            removedCounters: removedCountersSerialized,
        };
    }

    static generateLocalIdentifier() {
        return Math.floor(Math.random() * 10000);
    }

    static async createShoppingList(ring: HashRing): Promise<ShoppingList | null> {
        const id = crypto.randomUUID();
        const node = ring?.getResponsibleNode(id);
        const localIdentifier = Math.floor(Math.random() * 10000);
        const shoppingList = new ShoppingList(localIdentifier, id);

        const tries = 10;
        for (let i = 0; i < tries; i++) {
            try {
                // Switch api port from 8081 to dynamic once the backend is changed to reflect that
                const res = await fetch(`http://${node?.getHostname()}:${node?.getHttpPort()}/api/cart/${id}`, {
                    method: "PUT",
                    headers: {
                        "Content-Type": "application/json",
                    },
                    body: JSON.stringify({ "type": "shoppingList", ...shoppingList.toSerializable(false) }),
                });

                if (res.ok) {
                    return await res.json();
                }
            } catch (e) {
                console.error(e);
                continue;
            }
        }

        return shoppingList;
    }

    clone() {
        const cloned = new ShoppingList(this.localIdentifier, this.id);
        
        cloned.setItems(this.items.clone());
        cloned.setRemovedCounters(this.removedCounters);
        return cloned;
    }

    /**
     * list: {
      "id": "f1aef91a-0469-4869-beee-3e7658fd1e98",
      "localIdentifier": 50,
      "items": {
          "localIdentifier": 50,
          "values": {},
          "keys": {
              "localIdentifier": 50,
              "values": {},
              "dotContext": {
                  "dots": {}
              }
          }
      },
      "removedCounters": {}
  }
     */
    static fromDatabase(list) {
        try {
            const cloned = new ShoppingList(list.localIdentifier, list.id);
            cloned.setItems(AWMap.fromDatabase(list.items, list.localIdentifier) as AWMap<string, ShoppingListItem>);
            cloned.getItems().setLocalIdentifier(list.localIdentifier);
            
            if(Object.entries(list.removedCounters).length > 0) {
                const map = new Map();

                Object.entries(list.removedCounters).forEach(([key, value]) => {
                    map.set(key, value);
                });

                cloned.setRemovedCounters(map);
            } else cloned.setRemovedCounters(new Map());
            
            return cloned;
        } catch (error) {
            console.error(e);
        }
    }
}
